lib/Dialect/Torch/Transforms/GlobalizeObjectGraph.cpp

```diff
@@ -0,0 +1,338 @@
+LogicalResult ObjectGraphGlobalizer::globalizeObjectGraph() {
+  // We require there to be a unique root !torch.nn.Module.
+  FailureOr<NnModuleOp> maybeRootNnModule = findRootNnModule();
+  if (failed(maybeRootNnModule))
+    return failure();
+  NnModuleOp rootNnModule = *maybeRootNnModule;
+  if (!rootNnModule)
+    return module.emitError()
+           << "module does not contain a root torch.nn_module";
+
+  // We require one instance of each class. That is, there is a single
+  // torch.nn_module for each torch.class_type.
+  if (failed(checkSingleInstanceOfEachClass()))
+    return failure();
+
+  for (NnModuleOp nnModule : module.getOps<NnModuleOp>()) {
+    auto classType = symbolTable.lookup<ClassTypeOp>(nnModule.getClassName());
+    for (auto slot : nnModule.getOps<SlotOp>()) {
+      AttrOfClass attrOfClass = {classType, slot.name()};
+      slotInitialValues[attrOfClass] = slot.value();
+      slotInitialValuesInverseMap[slot.value()].push_back(attrOfClass);
+    }
+  }
+
+  // Recursively traverse the class hierarchy, globalizing slots and
+  // tracking linkage names for methods.
+  auto rootClassType =
+      symbolTable.lookup<ClassTypeOp>(rootNnModule.getClassName());
+  if (failed(recursivelyTraverseClassType(rootClassType)))
+    return failure();
+
+  // Move all slot initial values into an initializer func.
+  createInitializerFunc();
+
+  // Rewrite torch.prim.GetAttr/torch.prim.SetAttr/torch.prim.CallMethod.
+  if (failed(rewriteMethods()))
+    return failure();
+
+  // Now that all we have finished converting to the new form, remove
+  // the original object graph.
+  removeObjectGraph();
+
+  return success();
+}
+
+FailureOr<NnModuleOp> ObjectGraphGlobalizer::findRootNnModule() {
+  NnModuleOp rootNnModule;
+  for (NnModuleOp op : module.getOps<NnModuleOp>()) {
+    if (!op.use_empty())
+      continue;
+    if (rootNnModule) {
+      op.emitError()
+          .append("found more than one root module (module that is not a "
+                  "child of any other module)")
+          .attachNote(rootNnModule.getLoc())
+          .append("see other root module here");
+      return failure();
+    }
+    rootNnModule = op;
+  }
+  return rootNnModule;
+}
+
+LogicalResult ObjectGraphGlobalizer::checkSingleInstanceOfEachClass() {
+  llvm::MapVector</*ClassTypeOp*/ Operation *, std::vector<NnModuleOp>>
+      classInstances;
+  for (NnModuleOp op : module.getOps<NnModuleOp>()) {
+    auto classType = symbolTable.lookup<ClassTypeOp>(op.getClassName());
+    classInstances[classType].push_back(op);
+  }
+  for (auto &p : classInstances) {
+    ClassTypeOp classType = cast<ClassTypeOp>(p.first);
+    ArrayRef<NnModuleOp> instances = p.second;
+    if (instances.size() > 1) {
+      // TODO: Improve this diagnostic based on user use cases.
+      // This is a user-facing diagnostic that enforces key invariants to
+      // our TorchScript subset.
+      auto diag = classType.emitError("class type has more than one instance");
```

Comment:
Maybe just change it now to something like:

"class type has more than one instance: the current TorchScript supported subset only allows single instances"